server/src/main/java/org/elasticsearch/common/file/AbstractFileWatchingService.java

```diff
@@ -302,11 +302,12 @@ final WatchKey enableDirectoryWatcher(WatchKey previousKey, Path settingsDir) th
     void processSettingsOnServiceStartAndNotifyListeners() throws InterruptedException {
         try {
             processFileOnServiceStart();
-            for (var listener : eventListeners) {
-                listener.watchedFileChanged();
-            }
         } catch (IOException | ExecutionException e) {
-            logger.error(() -> "Error processing watched file: " + watchedFile(), e);
+            onProcessFileChangesException(e);
+            return;
+        }
+        for (var listener : eventListeners) {
```

Comment:
Quick check here: previously this loop was wrapped in the try-catch block, meaning that if an event listener threw IO or Execution we'd still continue. Feels like that's desirable -- I don't think we currently attach listeners in prod code (don't quote me on that though) so it's more of a hypothetical.

Same for `processSettingsAndNotifyListeners`. WDYT?

Reply:
Both `IOException` and `ExecutionException` are checked exceptions, so that's actually not possible. I pulled that loop out of the try-catch block partly for clarity.